### PR TITLE
Testing teal

### PR DESCRIPTION
**Pull request by robert-zaremba**
+ updated guide and tests

### CODE REVIEW

**Comment by robert-zaremba** [comment]
The goal of changing tests was to simplify it and to make it reference in our guide.

File: packages/runtime/test/integration/basic-teal.ts [hunk 1]
```diff
@@ -12,76 +12,63 @@ import { expectRuntimeError } from "../helpers/runtime-errors";
...
+  }
 
-  it("should send algo's from john to bob if stateless teal logic is correct", function () {
+  it("should send algo's from john to bob if delegated logic check passes", function () {
     // check initial balance
     assert.equal(john.balance(), initialJohnHolding);
     assert.equal(bob.balance(), initialBobHolding);
-    // get logic signature
+
+    // get delegated logic signature
```
Comment (ratik21) [inline packages/runtime/test/integration/basic-teal.ts:44]:
```suggestion
    // make delegated logic signature
```
I am not sure if "get" is the right word here. As we create (delegate) the logic signature at line 46.

File: packages/runtime/test/integration/escrow-account.ts [hunk 2]
```diff
@@ -14,116 +13,113 @@ import { johnAccount } from "../mocks/account";
 const minBalance = BigInt(ALGORAND_ACCOUNT_MIN_BALANCE + 1000); // 1000 to cover fee
 const initialEscrowHolding = minBalance + BigInt(1000e6);
 const initialJohnHolding = minBalance + 500n;
+const fee = 1000;
 
-describe("Algorand Stateless Smart Contracts (Contract Account Mode) - Escrow Account Example", function () {
+// The following test cases are used in Testiong TEAL guide,
```
Comment (ratik21) [inline packages/runtime/test/integration/escrow-account.ts:18]:
```suggestion
// The following test cases are used in Testing TEAL guide,
```

File: docs/guide/testing-teal.md [hunk 3]
```diff
@@ -4,60 +4,87 @@ layout: splash
 
 # Testing TEAL
 
-[TEAL](https://developer.algorand.org/docs/reference/teal/specification/) is a bytecode based stack language that executes inside Algorand transactions to check the parameters of the transaction and approve the transaction as if by a signature. `@algo-builder/runtime` provides a lightweight runtime and TEAL interpreter to test Algorand Smart Contracts.
+[TEAL](https://developer.algorand.org/docs/reference/teal/specification/) is a stack based language that executes inside Algorand transactions to program logic signatures and smart contracts. `@algo-builder/runtime` provides a JavaScript, lightweight runtime and TEAL interpreter to test Algorand transactions, ASA and Smart Contracts.
```
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply docs/guide/testing-teal.md:7]: OK, will update it.

File: packages/runtime/test/integration/escrow-account.ts [hunk 4]
```diff
@@ -14,116 +13,113 @@ import { johnAccount } from "../mocks/account";
...
   });
 
   it("should reject transaction if Fee > 10000", function () {
-    const invalidParams = Object.assign({}, txnParams);
-    invalidParams.payFlags = { totalFee: 12000 };
-
-    // execute transaction (should fail as fee is 12000)
     expectRuntimeError(
-      () => runtime.executeTx(invalidParams),
+      () => runtime.executeTx({ ...paymentTxParams, payFlags: { totalFee: 100000 } }),
```
Comment (ratik21) [inline packages/runtime/test/integration/escrow-account.ts:90]:
```suggestion
      () => runtime.executeTx({ ...paymentTxParams, payFlags: { totalFee: 0.1e6 } }),
```

File: packages/runtime/test/integration/escrow-account.ts [hunk 5]
```diff
@@ -14,116 +13,113 @@ import { johnAccount } from "../mocks/account";
...
-    amountMicroAlgos: initialEscrowHolding,
-    payFlags: { totalFee: 1000 }
-  };
-
-  let runtime: Runtime;
-  let escrow: AccountStore;
+  const runtime = new Runtime([john, admin]); // setup runtime
+
+  // we can't load teal code and create an escrow, because in this test we are loading a
+  // a fixture environment, which happens in `beforeAll`. So, consequently we need to move
```
Comment (ratik21) [inline packages/runtime/test/integration/escrow-account.ts:30]:
```suggestion
  // fixture environment, which happens in `beforeAll`. So, consequently we need to move
```

File: docs/guide/testing-teal.md [hunk 6]
```diff
@@ -4,60 +4,110 @@ layout: splash
...
+├── scripts
+│   ├── deploy.ts
+│   ├── run.ts
+├── test
+│   ├── JS test files
+│   ├── .mocharc.json (optional)
+├── package.json
+```
+
+All our test files should be stored in `test` directory. You can import `algob` library as well as `scripts`.
```
Comment (ratik21) [inline docs/guide/testing-teal.md:52]: Could you please explain "You can import `algob` library as well as `scripts`" ? It doesn't sound very clear.

File: docs/guide/testing-teal.md [hunk 7]
```diff
@@ -4,60 +4,110 @@ layout: splash
...
+Tests are typically done using [Mocha](https://mochajs.org/) framework, while assertions using [Chai](https://www.chaijs.com/) a BDD / TDD assertion library. Your test file is usually organized as follows:
+
+```
+describe("use-case", function() {
+  let variabl1;
```
Comment (ratik21) [inline docs/guide/testing-teal.md:57]:
```suggestion
  let variable1;
```

File: docs/guide/testing-teal.md [hunk 8]
```diff
@@ -70,137 +120,172 @@ See one of our examples for more details (eg: `examples/crowdfunding/test`).
...
-  ```js
-  it("should reject transaction if amount > 100", async function () {
-    const invalidParams = Object.assign({}, txnParams);
-    invalidParams.amountMicroAlgos = 500;
-
-    // execute transaction (should fail as amount = 500)
-    await expectRuntimeErrorAsync(
-      async () => await runtime.executeTx(invalidParams),
-      ERRORS.TEAL.REJECTED_BY_LOGIC
+  In the first test above, we fund the escrow using the admin account. John already has an initial balance set - we initialized runtime with Johns' account. In the second test we execute payment transaction from escrow to john and validate that the balances are correct.
```
Comment (ratik21) [inline docs/guide/testing-teal.md:187]:
```suggestion
  In the first test above, we fund the escrow using the admin account. John already has an initial balance set - we initialized runtime with John's account. In the second test we execute payment transaction from escrow to john and validate that the balances are correct.
```

File: docs/guide/testing-teal.md [hunk 9]
```diff
@@ -70,137 +120,172 @@ See one of our examples for more details (eg: `examples/crowdfunding/test`).
...
-    ```
+    runtime.executeTx(txnParams);
+
+    syncAccounts();
+    assert.equal(john.balance(), initialJohnHolding - 100n - BigInt(fee));
+    assert.equal(bob.balance(), initialBobHolding + 100n);
+  });
+  ```
+
+  - In the next test, create a delegated signature which verification will fail. We check that the transfer was not done and the balances didn't change.
```
Comment (ratik21) [inline docs/guide/testing-teal.md:245]:
```suggestion
  - In the next test, create a delegated signature whose verification will fail. We check that the transfer was not done and the balances didn't change.
```

File: docs/guide/testing-teal.md [hunk 10]
```diff
@@ -275,9 +359,11 @@ Currently, `runtime` supports:
 
 ## Examples
 
-Teal files used for the below tests can be found in `/test/fixtures` in `runtime`.
+TEAL files used for the below tests can be found in `/test/fixtures` in [runtime](https://github.com/scale-it/algo-builder/tree/master/packages/runtime/test/fixtures) package.
+
++ [Boilerplate Stateless Teal](https://github.com/scale-it/algo-builder/blob/master/packages/runtime/test/integration/basic-teal.ts)
++ [Escrow Account Test](https://github.com/scale-it/algo-builder/blob/master/packages/runtime/test/integration/escrow-account.ts)
++ [Boilerplate Stateful Teal](https://github.com/scale-it/algo-builder/blob/master/packages/runtime/test/integration/stateful-counter.ts)
++ Complex TEAK test suite (Stateless + Stateful + Atomic transactions) - [Crowdfunding application](https://github.com/scale-it/algo-builder/tree/master/examples/crowdfunding/test)
```
Comment (ratik21) [inline docs/guide/testing-teal.md:367]:
```suggestion
+ Complex TEAL test suite (Stateless + Stateful + Atomic transactions) - [Crowdfunding application](https://github.com/scale-it/algo-builder/tree/master/examples/crowdfunding/test)
```

File: docs/guide/testing-teal.md (hunk 6 above)
Comment (robert-zaremba) [reply docs/guide/testing-teal.md:52]: I mean that in tests you can import `algob` as a library and functions from scripts.

Comment (ratik21) [reply docs/guide/testing-teal.md]: ok

**Review by ratik21 — APPROVED** [review]

**Review by amityadav0 — APPROVED** [review]

**Pull request merged**